util/include/RawMemoryPool.hpp

```diff
@@ -0,0 +1,114 @@
+// initialization, a pre-defined number (minChunksNum_) of raw memory chunks of size chunkSize_ gets allocated.
+// In case there is no available chunk, a new one gets dynamically allocated - up to a maximum number (maxChunksNum_).
+// No support for different memory sizes provided.
+
+namespace concordUtil {
+
+class RawMemoryPoolMetrics {
+ public:
+  RawMemoryPoolMetrics(Timers& timers)
+      : component_{"memoryPoolMetrics", std::make_shared<concordMetrics::Aggregator>()},
+        availableChunks_{component_.RegisterAtomicGauge("availableChunks", 0)},
+        allocatedChunks_{component_.RegisterAtomicGauge("allocatedChunks", 0)},
+        timers_(timers) {
+    component_.Register();
+    const std::chrono::milliseconds period{100};
+    metricsTimer_ = timers_.add(period, Timers::Timer::RECURRING, [this](Timers::Handle h) { updateAggregator(); });
+  }
+
+  ~RawMemoryPoolMetrics() { timers_.cancel(metricsTimer_); }
+
+  void setAggregator(const std::shared_ptr<concordMetrics::Aggregator>& aggregator) {
+    component_.SetAggregator(aggregator);
+  }
+
+  void increaseAvailableChunksNum() { availableChunks_++; }
+  void decreaseAvailableChunksNum() { availableChunks_--; }
+  void increaseAllocatedChunksNum() { allocatedChunks_++; }
+  void decreaseAllocatedChunksNum() { allocatedChunks_--; }
+
+  void updateAggregator() { component_.UpdateAggregator(); }
+
+ private:
+  concordMetrics::Component component_;
+  concordMetrics::AtomicGaugeHandle availableChunks_;
+  concordMetrics::AtomicGaugeHandle allocatedChunks_;
+  Timers& timers_;
+  Timers::Handle metricsTimer_;
+};
+
+class RawMemoryPool {
+ public:
+  RawMemoryPool(uint32_t chunkSize, Timers& timers);
+  virtual ~RawMemoryPool();
+
+  void allocatePool(int32_t initialChunksNum, int32_t maxChunksNum);
+  int32_t getNumOfAvailableChunks() const { return numOfAvailableChunks_; }
+  int32_t getNumOfAllocatedChunks() const { return numOfAllocatedChunks_; }
+  bool isPoolFull() const { return numOfAllocatedChunks_ == maxChunksNum_; }
+  char* getChunk();
+  void returnChunk(char*);
+
+  RawMemoryPool(const RawMemoryPool&) = delete;
+  RawMemoryPool& operator=(const RawMemoryPool&) = delete;
+  RawMemoryPool(RawMemoryPool&&) = delete;
+  RawMemoryPool&& operator=(RawMemoryPool&&) = delete;
+
+  void setAggregator(const std::shared_ptr<concordMetrics::Aggregator>& aggregator) {
+    metrics_.setAggregator(aggregator);
+  }
+
+  static logging::Logger& logger() {
+    static logging::Logger logger_ = logging::getLogger("concord.memory.pool");
+    return logger_;
+  }
+
+ private:
+  bool isPoolPruningRequired();
+  void increaseNumOfAllocatedChunks();
+  void increaseNumOfAvailableChunks();
+  void decreaseNumOfAllocatedChunks();
+  void decreaseNumOfAvailableChunks();
+  char* allocateChunk();
+  void returnChunkToThePool(char* chunk);
+  void deleteChunk(char*& chunk);
+
+ private:
+  const uint32_t WAIT_TIMEOUT_MILLI = 5;
```

Comment:
Feels too short on the assumption that only the preprocessing takes 40ms.

Reply:
I think it's OK because we are not waiting for any specific request to complete. There are many requests on the fly - we use the same pool for all PreProcessor requests and any moment one of them could complete.